File: src/agentskills/bootstrap.py
#!/usr/bin/env python3
"""Install one or more skills into a project's .agents/skills directory."""

from __future__ import annotations

import argparse
import shutil
import subprocess
import sys
from pathlib import Path

DEFAULT_CACHE_DIR = Path.home() / ".cache" / "agentskills" / "repos"
PROJECT_SKILLS_DIR = Path(".agents") / "skills"


def run_command(
    cmd: list[str],
    cwd: Path | None = None,
    check: bool = True,
) -> subprocess.CompletedProcess[str]:
    """Run a shell command, raising on failure when check is True."""
    result = subprocess.run(
        cmd,
        cwd=str(cwd) if cwd else None,
        check=False,
        text=True,
        capture_output=True,
    )
    if check and result.returncode != 0:
        joined = " ".join(cmd)
        raise RuntimeError(f"Command failed ({joined}): {result.stderr.strip()}")
    return result


def parse_skill_list(raw_values: list[str]) -> list[str]:
    """Parse comma-separated and repeated skill names into a unique list."""
    parsed: list[str] = []
    for raw in raw_values:
        for part in raw.split(","):
            skill = part.strip()
            if skill and skill not in parsed:
                parsed.append(skill)
    return parsed


def repo_cache_name(repo_url: str) -> str:
    """Derive a filesystem-safe cache directory name from a repo URL."""
    trimmed = repo_url.rstrip("/")
    name = trimmed.rsplit("/", 1)[-1]
    if name.endswith(".git"):
        name = name[:-4]
    safe = "".join(
        ch if (ch.isalnum() or ch in ("-", "_", ".")) else "-" for ch in name
    )
    return safe or "skills-repo"


def clone_or_update_repo(repo_url: str, ref: str, cache_dir: Path) -> Path:
    """Clone or fetch+checkout a repo into the local cache."""
    cache_dir.mkdir(parents=True, exist_ok=True)
    repo_path = cache_dir / repo_cache_name(repo_url)

    if not repo_path.exists():
        run_command(["git", "clone", repo_url, str(repo_path)])
    else:
        run_command(
            ["git", "fetch", "--all", "--tags", "--prune"],
            cwd=repo_path,
        )

    run_command(["git", "checkout", ref], cwd=repo_path)
    pull_result = run_command(["git", "pull", "--ff-only"], cwd=repo_path, check=False)
    not_on_branch = "not currently on a branch" not in pull_result.stderr.lower()
    if pull_result.returncode != 0 and not_on_branch:
        print(
            f"warning: pull skipped: {pull_result.stderr.strip()}",
            file=sys.stderr,
        )
    return repo_path


def discover_available_skills(skills_root: Path) -> list[str]:
    """Return names of subdirectories containing a SKILL.md."""
    if not skills_root.exists():
        return []
    return [
        path.name
        for path in sorted(skills_root.iterdir())
        if path.is_dir() and (path / "SKILL.md").exists()
    ]


def remove_path(path: Path) -> None:
    """Remove a file, symlink, or directory tree."""
    if path.is_symlink() or path.is_file():
        path.unlink()
    elif path.exists():
        shutil.rmtree(path)


def install_skill(
    skill_name: str,
    source_root: Path,
    destination_root: Path,
    mode: str,
    force: bool,
) -> Path:
    """Copy or symlink a single skill into the destination."""
    source_skill = source_root / skill_name
    if not source_skill.exists():
        raise RuntimeError(f"Skill not found in source repo: {source_skill}")

    destination_skill = destination_root / skill_name
    if destination_skill.exists() or destination_skill.is_symlink():
        if not force:
            raise RuntimeError(
                f"Destination already exists: {destination_skill}"
                " (use --force to replace)"
            )
        remove_path(destination_skill)

    if mode == "copy":
        shutil.copytree(source_skill, destination_skill)
    else:
        destination_skill.symlink_to(source_skill.resolve(), target_is_directory=True)
    return destination_skill


def parse_args() -> argparse.Namespace:
    """Parse CLI arguments for the bootstrap command."""
    parser = argparse.ArgumentParser(
        description="Bootstrap agent skills into a project.",
    )
    parser.add_argument("--project", required=True, help="Target project path.")
    parser.add_argument(
        "--skill",
        action="append",
        default=[],
        help="Skill name(s). Repeat or comma-separate. Default: all.",
    )
    parser.add_argument(
        "--mode",
        choices=["copy", "symlink"],
        default="copy",
        help="Install mode.",
    )
    parser.add_argument(
        "--force",
        action="store_true",
        help="Replace destination skill if it already exists.",
    )
    parser.add_argument("--repo-url", help="Git URL to clone/pull from.")
    parser.add_argument(
        "--repo-path",
        help="Use a local repo path instead of cloning.",
    )
    parser.add_argument(
        "--ref",
        default="main",
        help="Git branch or tag when --repo-url is used.",
    )
    parser.add_argument(
        "--cache-dir",
        default=str(DEFAULT_CACHE_DIR),
        help="Clone cache directory.",
    )
    return parser.parse_args()


def resolve_repo_root(args: argparse.Namespace) -> Path:
    """Determine the repo root from CLI arguments."""
    if args.repo_path and args.repo_url:
        raise RuntimeError("Use either --repo-path or --repo-url, not both.")

    if args.repo_path:
        repo_root = Path(args.repo_path).expanduser().resolve()
        if not repo_root.exists():
            raise RuntimeError(f"Repo path does not exist: {repo_root}")
        return repo_root

    if args.repo_url:
        return clone_or_update_repo(
            repo_url=args.repo_url,
            ref=args.ref,
            cache_dir=Path(args.cache_dir).expanduser().resolve(),
        )

    return Path.cwd().resolve()


def main() -> int:
    """Bootstrap skills into a target project."""
    try:
        args = parse_args()
        repo_root = resolve_repo_root(args)

        skills_root = repo_root / "skills"
        available = discover_available_skills(skills_root)
        if not available:
            raise RuntimeError(f"No skills found under: {skills_root}")

        requested = parse_skill_list(args.skill)
        selected = requested if requested else available
        unknown = [name for name in selected if name not in available]
        if unknown:
            raise RuntimeError(
                f"Unknown skill(s): {', '.join(unknown)}. "
                f"Available: {', '.join(available)}"
            )

        project_root = Path(args.project).expanduser().resolve()
        project_root.mkdir(parents=True, exist_ok=True)
        destination_root = project_root / PROJECT_SKILLS_DIR
        destination_root.mkdir(parents=True, exist_ok=True)

        installed: list[Path] = []
        for skill_name in selected:
            installed.append(
                install_skill(
                    skill_name=skill_name,
                    source_root=skills_root,
                    destination_root=destination_root,
                    mode=args.mode,
                    force=args.force,
                )
            )

        print(f"repo: {repo_root}")
        print(f"project: {project_root}")
        print(f"mode: {args.mode}")
        print("installed:")
        for path in installed:
            print(f"  - {path}")
        print("next: run optional harness links with `agentskills link` if needed.")
        return 0
    except Exception as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/agentskills/cmd/bootstrap.go
package cmd

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/jwa91/agentskills/tools/agentskills/internal/skillrepo"
)

const projectSkillsRel = ".agents/skills"

// RunBootstrap is the entrypoint for `agentskills bootstrap`.
func RunBootstrap(args []string) error {
	fs := SubFlagSet("bootstrap", "install one or more skills into a project's .agents/skills directory")
	var (
		project    = fs.String("project", ".", "target project path")
		mode       = fs.String("mode", "copy", "install mode: copy (safest) or symlink")
		force      = fs.Bool("force", false, "replace destination skill if it already exists")
		repoURL    = fs.String("repo-url", "", "git URL to clone/pull from (defaults to $AGENTSKILLS_REPO or upstream)")
		repoPath   = fs.String("repo-path", "", "use a local repo path instead of cloning")
		ref        = fs.String("ref", "", "git branch or tag when --repo-url is used (defaults to $AGENTSKILLS_REF or 'main')")
		cacheDir   = fs.String("cache-dir", "", "clone cache directory")
		installAll = fs.Bool("all", false, "install all available skills without prompting")
	)
	var skillFlag []string
	fs.Var(newRepeatableList(&skillFlag), "skill", "skill name(s); repeat or comma-separate")

	if err := ParseFlags(fs, args); err != nil {
		return err
	}

	if *repoURL != "" && *repoPath != "" {
		return fmt.Errorf("use either --repo-path or --repo-url, not both")
	}
	if *mode != "copy" && *mode != "symlink" {
		return fmt.Errorf("--mode must be 'copy' or 'symlink' (got %q)", *mode)
	}

	resolvedCache := *cacheDir
	if resolvedCache == "" {
		c, err := skillrepo.DefaultCacheDir()
		if err != nil {
			return err
		}
		resolvedCache = c
	}

	repoRoot, err := resolveRepoRoot(*repoURL, *repoPath, *ref, resolvedCache)
	if err != nil {
		return err
	}

	skillsRoot := filepath.Join(repoRoot, "skills")
	available := skillrepo.DiscoverAll(skillsRoot)
	if len(available) == 0 {
		return fmt.Errorf("no skills found under: %s", skillsRoot)
	}

	requested := SplitCommaList(skillFlag)
	var selected []string
	switch {
	case len(requested) > 0:
		selected = requested
	case *installAll:
		selected = available
	default:
		picked, err := pickSkillsInteractive(skillsRoot)
		if err != nil {
			return err
		}
		selected = picked
	}

	availSet := map[string]bool{}
	for _, n := range available {
		availSet[n] = true
	}
	var unknown []string
	for _, n := range selected {
		if !availSet[n] {
			unknown = append(unknown, n)
		}
	}
	if len(unknown) > 0 {
		return fmt.Errorf("unknown skill(s): %s. available: %s",
			strings.Join(unknown, ", "), strings.Join(available, ", "))
	}

	absProject, err := filepath.Abs(expandHome(*project))
	if err != nil {
		return err
	}
	if err := os.MkdirAll(absProject, 0o755); err != nil {
		return err
	}
	destinationRoot := filepath.Join(absProject, projectSkillsRel)
	if err := os.MkdirAll(destinationRoot, 0o755); err != nil {
		return err
	}

	if *mode == "symlink" {
		fmt.Fprintln(os.Stderr, "warning: symlink mode creates links that resolve outside the project directory. This may cause permission prompts in Claude Code for every file read. Use --mode copy to avoid this.")
	}

	var installed []string
	for _, name := range selected {
		p, err := skillrepo.InstallSkill(name, skillsRoot, destinationRoot, *mode, *force)
		if err != nil {
			return err
		}
		installed = append(installed, p)
	}

	fmt.Printf("repo: %s\n", repoRoot)
	fmt.Printf("project: %s\n", absProject)
	fmt.Printf("mode: %s\n", *mode)
	fmt.Println("installed:")
	for _, p := range installed {
		fmt.Printf("  - %s\n", p)
	}
	fmt.Println("next: run optional harness links with `agentskills link` if needed.")
	return nil
}

func resolveRepoRoot(repoURL, repoPath, ref, cacheDir string) (string, error) {
	if repoPath != "" {
		abs, err := filepath.Abs(expandHome(repoPath))
		if err != nil {
			return "", err
		}
		if _, err := os.Stat(abs); err != nil {
			return "", fmt.Errorf("repo path does not exist: %s", abs)
		}
		return abs, nil
	}

	url := repoURL
	if url == "" {
		// Only fall through to clone if user explicitly asked (via --repo-url
		// or env). Otherwise we assume the user is running from inside a
		// clone of jwa91/agentskills and try the source tree.
		if env := os.Getenv("AGENTSKILLS_REPO"); env != "" {
			url = env
		}
	}

	if url == "" {
		if root, ok := findLocalRepoRoot(); ok {
			return root, nil
		}
		// Fall back to default upstream URL.
		url = DefaultRepoURL()
	}

	r := ref
	if r == "" {
		r = DefaultRef()
	}
	return skillrepo.CloneOrUpdate(url, r, cacheDir)
}

// findLocalRepoRoot walks up from cwd looking for a skills/ directory.
// This matches the Python `REPO_ROOT` heuristic: when running from within
// the agentskills checkout, prefer the local tree over a network clone.
func findLocalRepoRoot() (string, bool) {
	cwd, err := os.Getwd()
	if err != nil {
		return "", false
	}
	dir := cwd
	for i := 0; i < 8; i++ {
		if fi, err := os.Stat(filepath.Join(dir, "skills")); err == nil && fi.IsDir() {
			return dir, true
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			break
		}
		dir = parent
	}
	return "", false
}

func expandHome(p string) string {
	if strings.HasPrefix(p, "~") {
		if home, err := os.UserHomeDir(); err == nil {
			return filepath.Join(home, strings.TrimPrefix(p, "~"))
		}
	}
	return p
}

// pickSkillsInteractive is a tiny stdlib-only TTY picker. The user types
// space- or comma-separated indices, optionally a/A to toggle all.
// It deliberately stays minimal — the briefing tells us the picker should
// be a deliberate choice, not a default.
func pickSkillsInteractive(skillsRoot string) ([]string, error) {
	if !isStdinTTY() {
		return nil, fmt.Errorf("no skills specified. Use --skill <name>, --all, or `agentskills list` to see available skills")
	}
	own, curated := skillrepo.Categorize(skillsRoot)
	if len(own)+len(curated) == 0 {
		return nil, fmt.Errorf("no skills found under: %s", skillsRoot)
	}
	type row struct {
		name    string
		curated bool
	}
	var rows []row
	for _, n := range own {
		rows = append(rows, row{n, false})
	}
	for _, n := range curated {
		rows = append(rows, row{n, true})
	}

	fmt.Fprintln(os.Stderr, "Available skills:")
	for i, r := range rows {
		label := r.name
		if r.curated {
			label += " (curated)"
		}
		fmt.Fprintf(os.Stderr, "  %2d) %s\n", i+1, label)
	}
	fmt.Fprintln(os.Stderr, "Enter numbers separated by space or comma (or 'a' for all):")
	fmt.Fprint(os.Stderr, "> ")

	reader := bufio.NewReader(os.Stdin)
	line, err := reader.ReadString('\n')
	if err != nil {
		return nil, fmt.Errorf("read selection: %w", err)
	}
	line = strings.TrimSpace(line)
	if line == "" {
		return nil, fmt.Errorf("no selection")
	}
	if line == "a" || line == "A" {
		out := make([]string, 0, len(rows))
		for _, r := range rows {
			out = append(out, r.name)
		}
		return out, nil
	}

	fields := strings.FieldsFunc(line, func(r rune) bool {
		return r == ',' || r == ' ' || r == '\t'
	})
	seen := map[string]bool{}
	var out []string
	for _, f := range fields {
		idx, err := strconv.Atoi(strings.TrimSpace(f))
		if err != nil {
			return nil, fmt.Errorf("invalid selection %q", f)
		}
		if idx < 1 || idx > len(rows) {
			return nil, fmt.Errorf("selection out of range: %d", idx)
		}
		name := rows[idx-1].name
		if !seen[name] {
			seen[name] = true
			out = append(out, name)
		}
	}
	if len(out) == 0 {
		return nil, fmt.Errorf("no skills selected")
	}
	return out, nil
}

func isStdinTTY() bool {
	fi, err := os.Stdin.Stat()
	if err != nil {
		return false
	}
	return (fi.Mode() & os.ModeCharDevice) != 0
}
